onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -114,10 +114,114 @@ static Status UnpackTensorWithRawData(const void* raw_data, size_t raw_data_leng
+      "External data type cannot be UNDEFINED or STRING.");
+
+  std::unique_ptr<onnxruntime::ExternalDataInfo> external_data_info;
+  ORT_RETURN_IF_ERROR(onnxruntime::ExternalDataInfo::Create(tensor_proto.external_data(), external_data_info));
+
+  if (tensor_proto_dir != nullptr) {
+    external_file_path = onnxruntime::ConcatPathComponent<ORTCHAR_T>(tensor_proto_dir, external_data_info->GetRelPath());
+  } else {
+    external_file_path = external_data_info->GetRelPath();
+  }
+
+  file_offset = external_data_info->GetOffset();
+
+  ORT_RETURN_IF_ERROR(onnxruntime::utils::GetSizeInBytesFromTensorProto<0>(
+      tensor_proto, &tensor_data_length));
+  const size_t external_data_length = external_data_info->GetLength();
+
+  ORT_RETURN_IF_NOT(
+      external_data_length == 0 ||
+          external_data_length == tensor_data_length,
+      "TensorProto external data size mismatch. ",
+      "Computed size: ", tensor_data_length,
+      ", external_data.length: ", external_data_length);
+
+  return Status::OK();
+}
+
+// Read external data for tensor in unint8_t* form and return Status::OK() if the data is read successfully.
+// Uses the tensor_proto_dir to construct the full path for external data. If tensor_proto_dir == nullptr
+// then uses the current directory instead.
+// This function does not unpack string_data of an initializer tensor
+static Status ReadExternalDataForTensor(const ONNX_NAMESPACE::TensorProto& tensor_proto,
+                                        const ORTCHAR_T* tensor_proto_dir,
+                                        std::unique_ptr<uint8_t[]>& unpacked_tensor,
+                                        size_t& tensor_data_length) {
+  std::basic_string<ORTCHAR_T> external_file_path;
+  onnxruntime::FileOffsetType file_offset;
+  GetExternalDataInfo(tensor_proto, tensor_proto_dir, external_file_path, file_offset, tensor_data_length);
+  unpacked_tensor.reset(new uint8_t[tensor_data_length]);
+  ORT_RETURN_IF_ERROR(onnxruntime::Env::Default().ReadFileIntoBuffer(
+      external_file_path.c_str(),
+      file_offset,
+      tensor_data_length,
+      gsl::make_span(reinterpret_cast<char*>(unpacked_tensor.get()), tensor_data_length)));
+
+  return Status::OK();
+}
 }  // namespace
 
 namespace onnxruntime {
 namespace utils {
+#if !defined(ORT_MINIMAL_BUILD)
+#define DEFINE_UNPACK_EXTERNAL_TENSOR(T)                                                                                   \
+  template <>                                                                                                              \
+  Status UnpackTensorWithExternalData(const ONNX_NAMESPACE::TensorProto& tensor,                                           \
+                                      const ORTCHAR_T* tensor_proto_dir, size_t expected_size,                             \
+                                      /*out*/ T* p_data) {                                                                 \
```

Comment:
nit: ptr for out data vs ref.

Reply:
using T* following the rest of the UnpackTensor calls... added nullptr check for p_data as that was missing here

Reply:
Our coding conventions say reference unless it can be nullptr. Even google these days says use a reference (what a time to be alive!).

https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs
> Input parameters should usually be values or const references, while non-optional output and input/output parameters should usually be references (which cannot be null)